02_bubble/pfannerj/main.go

```diff
@@ -0,0 +1,62 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var mainout io.Writer = os.Stdout
+
+func main() {
+	fmt.Fprintf(mainout, "Bubble sort starting...\n")
+	fmt.Fprintln(mainout, bubble([]int{3, 2, 1, 5}))
+	fmt.Fprintf(mainout, "Bubble sort completed...\n")
+
+	fmt.Fprintf(mainout, "Insertion sort starting...\n")
+	fmt.Fprintln(mainout, insertion([]int{3, 2, 5, 1}))
+	fmt.Fprintf(mainout, "Insertion sort completed...\n")
+}
+
+func bubble(s []int) []int {
+	sCopy := make([]int, len(s))
+	copy(sCopy, s)
+	var saveNbr int
+	swapped := false
+	i := 0
+	for {
+		if (i + 1) >= len(s) {
+			i = 0
+			if swapped {
+				swapped = false
+			} else {
+				break //Sort is complete if no values were 'swapped' during the previous iteration
+			}
+		}
+		if sCopy[i] > sCopy[i+1] {
+			saveNbr = sCopy[i]
+			sCopy[i], sCopy[i+1] = sCopy[i+1], saveNbr
```

Comment:
You don't actually need `saveNbr`, `sCopy[i], sCopy[i+1] = sCopy[i+1], sCopy[i]` works just as well, you won't lose either number. The reason for this is all arguments on the right are evaluated before being assigned.